AzureAD and SAML Doco (#1515)
FusionAuth/fusionauth-site: site/docs/v1/tech/identity-providers/samlv2/azure-ad.adoc

image:identity-providers/azure-ad-saml/configure-single-sign-on-saml[Configure Single-sign on saml,width=1200,role=shadowed]

AzureAD will present five steps to complete. However, we have a bit of a procedural dilemma. Before we can move forward, we need to make an Entity ID (our FusionAuth Idp, in this case) so AzureAD can generate a X509 certificate. However, before we make an Entity ID/IdP in FusionAuth, we need an X509 certificate. The below steps will solve this problem and complete our integration. In sum, we will be:

1. Creating a "dummy" X509 certificate and importing this into Key Master.
2. Making a SAML Idp in FusionAuth (Entity Id) with a dummy certificate selected.
3. Updating our AzureAD Enterprise application with values from our newly created Idp, thereby getting a valid X509 certificate.
4. Revisiting our SAML Idp in FusionAuth and updating with the correct X509 certificate from the above step.
5. Completing our integration.

If you already have a valid X509 certificate, you can complete step two above using this valid certificate and skip a few of the steps below.

With that discussed, let us begin!

== Import a "Dummy" X509 Certificate

> The first step is to import a "dummy" certificate into Key Master. You can generate any X509 certificate as it will be changed later. Once generated, under Key Master (`Settings > Key Master`) click `Import Certificate`.

Reviewer: Should you provide an example using openssl showing how to create a placeholder cert? That'd be nice.

image:identity-providers/azure-ad-saml/import-certificate[Import Certificate,width=1200,role=shadowed]

> Included is an example X509 certificate for sample use.

Reviewer: Ah, so they don't even need to create one, they can just use this one? Any reason why we wouldn't just tell folks to import this one (and not even mess with generation)?

Author: Good point!

Author: Was considering this, but thought that it would be a bit presumptious that they would use my certificate 😄

```
-----BEGIN CERTIFICATE-----
MIIC1DCCAj2gAwIBAgIBADANBgkqhkiG9w0BAQ0FADCBhjELMAkGA1UEBhMCdXMx
DzANBgNVBAgMBkRlbnZlcjETMBEGA1UECgwKRnVzaW9uQXV0aDEWMBQGA1UEAwwN
RnVzaW9uQXV0aC5pbzETMBEGA1UECwwKRnVzaW9uQXV0aDEkMCIGCSqGSIb3DQEJ
ARYVcmljaGFyZEBwaWVkcGlwZXIuY29tMB4XDTIyMDcxNjA1MDc1N1oXDTIzMDcx
NjA1MDc1N1owgYYxCzAJBgNVBAYTAnVzMQ8wDQYDVQQIDAZEZW52ZXIxEzARBgNV
BAoMCkZ1c2lvbkF1dGgxFjAUBgNVBAMMDUZ1c2lvbkF1dGguaW8xEzARBgNVBAsM
CkZ1c2lvbkF1dGgxJDAiBgkqhkiG9w0BCQEWFXJpY2hhcmRAcGllZHBpcGVyLmNv
bTCBnzANBgkqhkiG9w0BAQEFAAOBjQAwgYkCgYEA2FmZJHJFVEM9JwFd6Za87T0Z
MtIL5djSFC/TBVqhCx15eauNGAV/RoulESA6qsI4LNrbJ8uEYDQa9UXAZCc9yRMa
e/+E5XAApV4K06duo+vKon5L21YZ7HxzjUfL3bhLqKvpFWCQkNrH0rxgPCGwzh7N
T24sFcKwaVvBdknm9i8CAwEAAaNQME4wHQYDVR0OBBYEFLdhYvqAwTBCEQsSZdKj
UVR57CwDMB8GA1UdIwQYMBaAFLdhYvqAwTBCEQsSZdKjUVR57CwDMAwGA1UdEwQF
MAMBAf8wDQYJKoZIhvcNAQENBQADgYEAIpQags/uHj0dyFcCtRcPeVEDUqBPZaGO
M9kbFiGATlMNX4OiDvMUKs7Ag9109o0iLWPvLBQrMDn87fSy6+MUXZRS4asjzJCp
5aVWSevI85H6xS8WXxFr21etaqfiE88Lw86gK5O4iKtMBtCnWA5iUc2EJ0citQ0G
Pk8ybmMP1r8=
-----END CERTIFICATE-----
```

== Create a SAML v2 Identity Provider

> Returning to the FusionAuth Admin UI, navigate to `Settings > Identity Providers > Add`.

Reviewer: ```suggestion
In the FusionAuth administrative user interface, navigate to [breadcrumb]#Settings -> Identity Providers -> Add#.
```

The identity provider we will be adding is a `SAML V2` identity provider.

image:identity-providers/azure-ad-saml/idp-saml-configuration[IdP SAML configuration,width=1200,role=shadowed]

The following fields will need to be completed.

> - _Name_. This is user-defined.

Reviewer: ```suggestion
- [field]#Name#. This is user-defined, so pick something that describes this connection such as "Pied Piper Azure AD".
```

> - _Idp endpoint_. This value can be obtained from your AzureAD Application as demonstrated below.

Reviewer: ```suggestion
- [field]#IdP endpoint#. This value can be obtained from your Azure AD Application as demonstrated below.
```

Reviewer: I'm confused. What should I put in IdP endpoint?

Reviewer: Or are we going to edit and update this value once we have an entity Id to provide to Azure AD? If so, you should say that.

> image:identity-providers/azure-ad-saml/step-four-idp-endpoint[Step Four,width=1200,role=shadowed]

Reviewer: Need a better alt text here.

> 

Reviewer: ```suggestion
```

> - _Verification key_. For this value, please use the dummy certificate imported above. We will change this later.

Reviewer: ```suggestion
- [field]#Verification key#. For this value, use the dummy certificate imported above. You will change this later.
```

> Be sure to select `Debug Enabled`. This will ensure debug information is written to the `Event Log`. Remember to disable this option in production.

Reviewer: ```suggestion
Be sure to check `Debug Enabled`. This will ensure debug information is written to the Event Log, which you can see by navigating to [breadcrumb]#System -> Event Log#. 

Remember to disable this option when running in production.
```

[NOTE]
====
More information regarding the event log can be found below

> - link:https://site-local.fusionauth.io/docs/v1/tech/apis/event-logs#overview[API for Event Log]

Reviewer: Use relative links.

> - link:https://site-local.fusionauth.io/docs/v1/tech/admin-guide/troubleshooting#event-log[Admin UI and Troubleshooting]

Reviewer: Use relative links.

====

> Lastly, optionally configure the Relay State and Logout URL (not all integrations will require this).

Reviewer: Hmmm. I don't see either of these in the FusionAuth IdP config. What am I missing?

Author: Removed this irrelevant reference.

> Ensure you have enabled this IdP in your FusionAuth application.

Reviewer: ```suggestion
Then, navigate to [breadcrumb]#Applications# in the administrative user interface. 

Enable this IdP for the appropriate Application.
```

image:identity-providers/azure-ad-saml/enable-idp[Enable Idp On Application,width=1200,role=shadowed]

> Once saved, you may recieve a CORS warning. Be sure to follow the onscreen instructions and this link:/docs/v1/tech/identity-providers/samlv2/#cors-configuration[documentation].

Reviewer: ```suggestion
Once saved, you may receive a CORS warning. Be sure to follow the onscreen instructions and this link:/docs/v1/tech/identity-providers/samlv2/#cors-configuration[documentation].
```

=== Linking Options

> To note, if you are link:/docs/v1/tech/identity-providers/#linking-strategies[linking to a user on email], please configure the SAML Idp to find the email claim. This can be done by navigating to `Settings > Identity Providers > Your SAML Idp > Edit > Options Tab` and adjusting the `Email claim` field.

Reviewer: ```suggestion
If you are link:/docs/v1/tech/identity-providers/#linking-strategies[linking to a user on email], please configure the SAML Idp to find the email claim. This can be done by navigating to [breadcrumb]#Settings -> Identity Providers -> Your SAML IdP -> Edit# and then the [breadcrumb]#Options# tab. Adjust the [field]#Email claim# field.
```

If you are using an existing AzureAD Application this email claim location may vary. Below is a demonstration of the email claim found in a default AzureAD Application.

> 

Reviewer: I would put this as text so they can copy it.

I would also explain where they can find the email claim location. Seems important.

image:identity-providers/azure-ad-saml/email-claim[Setup custom email claim,width=1200,role=shadowed]

== Finish Setup of Azure AD Application

> With the Identity Provider setup complete, the next step is to input our newly created values into our AzureAD Application.

Reviewer: ```suggestion
With the FusionAuth Identity Provider set up, the next step is to input th newly created values into your Azure AD Application.
```

Navigate to your AzureAD application. Under step one, click on the pencil to edit.

image:identity-providers/azure-ad-saml/saml-5-steps-overview[SAML 5 steps overview,width=1200,role=shadowed]

> On this screen add the Entity ID and Reply URL. The Entity ID can be obtained by going to `Setting > Identity Providers > your SAML Idp > Magnifying Glass Icon > SAML v2 Integration details`. This value is synonymous with the Issuer value in FusionAuth.

Reviewer: What about the Reply URL?

Reviewer: ```suggestion
On this screen add the [field]#Entity ID# and [field]#Reply URL#. The Entity ID can be obtained by going to [breadcrumb]#Setting -> Identity Providers# and clicking on the Magnifying Glass in the row of your Identity Provider. This value is synonymous with the Issuer value in FusionAuth.
```

image:identity-providers/azure-ad-saml/integration-details[Integration Details,width=1200,role=shadowed]

== Download and Import X509 Certificate

> Previously, we created and used a "dummy" X509 certificate. At this point, we can change this. From the overview of your SAML application in AzureAD, under step three, you should find a Certificate (Base64) to download.

Reviewer: ```suggestion
Previously, you created and used a "dummy" X.509 certificate. At this point, you can change this to the real value. In the Azure AD SAML application overview, under step three, find a [field]#Certificate (Base64)#. You'll download and then upload this to FusionAuth.
```

image:identity-providers/azure-ad-saml/generated-certificate[Generated Certificate,width=1200,role=shadowed]

Next, import this certificate into Key Master (`Settings > Key Master > Import Certificate`). Leave the Key identifier property blank, as this will be autogenerated from the thumbprint in the existing certificate.

Update the certificate used on your Idp configuration (`Settings > Identity Providers > your SAML IDP > Verification Key`) with this new X509 certificate. You can also remove the "dummy" certificate from Key Master once complete.

[WARNING]
====
Leave the Key identifier property blank, as this will be autogenerated from the thumbprint of the existing certificate. Otherwise, FusionAuth will not be able to validate the signature of the AuthN response from AzureAD when using this integration.
====

== Assign Users

[NOTE]
====
If you are using an existing AzureAD application with assigned users, you may skip this step.
====

> Before we can complete our integration, assign AzureAD users to your application. Add a user by navigating to `Users and groups` and clicking on the `Add user/group` button. Follow onscreen instructions to select and add a user.

Reviewer: ```suggestion
Before we can complete our integration, assign Azure AD users to your Azure AD application. Add a user by navigating to [breadcrumb]#Users and groups# and clicking on the Add user/group` button. Follow the instructions to select and add a user.
```

image:identity-providers/azure-ad-saml/add-user-azure[Add a User to Azure,width=1200,role=shadowed]

== Complete the Login

image:identity-providers/azure-ad-saml/new-login-button[New Login With Azure-Button,width=1200,role=shadowed]

> At this point, what remains is to attempt a login. To test quickly, navigate to `Applications > Your Application > View (magnify icon) > OAuth2 & OpenID Connect Integration details > Login URLs`. Copy is this URL.

Reviewer: ```suggestion
At this point, what remains is to attempt a login. To test quickly, log into the FusionAuth administrative user inteface. Navigate to [breadcrumb]#Applications -> Your Application# Click the [breadrumb]#View# link (the magnify icon), then scroll to the [breadcrumb]#OAuth2 & OpenID Connect Integration details# section and look for the Login URL. Copy this.
```

> Open a new tab, paste in the login URL, and you should see your new SAML AzureAD IdP login option. If your integration fails, remember to review the Event Logs (`System > Event Log`)

Reviewer: ```suggestion
Open a new tab, paste in the login URL, and you should see your new SAML AzureAD IdP login button. 

Click on that button, log in with one of the Azure AD users you configured above, and you should arrive back at your redirect URL.

If your integration fails, remember to review the Event Logs (`System > Event Log`)
```

Did you show them how to configure the redirect URL in the FusionAuth application?

[NOTE]
====
More information regarding the Event Log can be found below:

> - link:https://site-local.fusionauth.io/docs/v1/tech/apis/event-logs#overview[API for Event Log]

Reviewer: relative links.

> - link:https://site-local.fusionauth.io/docs/v1/tech/admin-guide/troubleshooting#event-log[Admin UI and Troubleshooting]

Reviewer: Relative links.
